blackbird_python/blackbird/tests/test_program.py

```diff
@@ -0,0 +1,337 @@
+        )
+        res = bb.serialize()
+        expected = dedent(
+            """\
+            name prog
+            version 1.0
+
+            Dgate(alpha=0.43-0.543j, phi=0.54) | 0
+            """
+        )
+        assert res == expected
+
+    def test_serialize_operation_args_kwargs(self):
+        """Test serialization of an operation with args and kwargs raises an error"""
+        bb = BlackbirdProgram(name="prog", version=1.0)
+        bb._operations.append(
+            {
+                "op": "Dgate",
+                "modes": [0],
+                "args": [0.54],
+                "kwargs": {"alpha": 0.43 - 0.543j, "phi": 0.54},
+            }
+        )
+
+        with pytest.raises(
+            ValueError, match="contain both positional and keyword arguments"
+        ):
+            bb.serialize()
+
+    def test_serialize_operation_multimode(self):
+        """Test serialization of an operation on multiple modes"""
+        bb = BlackbirdProgram(name="prog", version=1.0)
+        bb._operations.append({"op": "Vac", "modes": [0, 1, 2]})
+        res = bb.serialize()
+        expected = dedent(
+            """\
+            name prog
+            version 1.0
+
+            Vac | [0, 1, 2]
+            """
+        )
+        assert res == expected
+
+    def test_serialize_operation_array_arg(self):
+        """Test serialization of an operation with an array arg"""
+        bb = BlackbirdProgram(name="prog", version=1.0)
+        U = np.identity(2)
+
+        bb._operations.append(
+            {"op": "Interferometer", "modes": [0], "args": [U], "kwargs": {}}
+        )
+
+        res = bb.serialize()
+        expected = dedent(
+            """\
+            name prog
+            version 1.0
+
+            int array A0[2, 2] =
+                1, 0
+                0, 1
+
+            Interferometer(A0) | 0
+            """
+        )
+        assert res == expected
+
+    def test_serialize_operation_array_kwarg(self):
+        """Test serialization of an operation with an array kwarg"""
+        bb = BlackbirdProgram(name="prog", version=1.0)
+        U = np.identity(2)
+
+        bb._operations.append(
+            {"op": "Interferometer", "modes": [0], "args": [], "kwargs": {"U": U}}
+        )
+
+        res = bb.serialize()
+        expected = dedent(
+            """\
+            name prog
+            version 1.0
+
+            int array A0[2, 2] =
+                1, 0
+                0, 1
+
+            Interferometer(U=A0) | 0
+            """
+        )
+        assert res == expected
+
+    def test_serialize_operation_string_arg(self):
+        """Test serialization of an operation with a string arg"""
+        bb = BlackbirdProgram(name="prog", version=1.0)
+        bb._operations.append({"op": "Dgate", "modes": [0], "args": ["hi"], "kwargs": {}})
+        res = bb.serialize()
+        expected = dedent(
+            """\
+            name prog
+            version 1.0
+
+            Dgate("hi") | 0
```

Comment:
lol is there not a native operation which takes string arguments?

Reply:
Nope haha